src/main/java/uk/gov/hmcts/reform/et/syaapi/service/CaseDocumentService.java

```diff
@@ -0,0 +1,211 @@
+    private static final Pattern FILE_NAME_PATTERN = Pattern.compile(FILE_NAME_REGEX_PATTERN);
+    private static final String HTTPS_URL_REGEX_PATTERN =
+        "^(http://www\\.|https://www\\.|http://|https://)?[\\w\\-]+(\\.[a-z]{2,5})?(:\\d{1,5})?(/.*)?$";
+    private static final Pattern HTTPS_URL_PATTERN = Pattern.compile(HTTPS_URL_REGEX_PATTERN);
+    private static final String UPLOAD_FILE_EXCEPTION_MESSAGE = "Document management failed uploading file: ";
+    private static final String VALIDATE_FILE_EXCEPTION_MESSAGE = "File does not pass validation";
+    public static final int MAX_API_RETRIES = 3;
+    private final RestTemplate restTemplate;
+    private final AuthTokenGenerator authTokenGenerator;
+    private final String caseDocApiUrl;
+
+    /**
+     * Default constructor with injected parameters.
+     *
+     * @param restTemplate       the {@link RestTemplate} to be used to connect with the Case Document API
+     * @param authTokenGenerator the {@link AuthTokenGenerator} used to generate tokens for communicating with the
+     *                           Case Document API
+     * @param caseDocApiUrl      the URL to call the Case Document API
+     */
+    public CaseDocumentService(RestTemplate restTemplate,
+                               AuthTokenGenerator authTokenGenerator,
+                               @Value("${case_document_am.url}/cases/documents}")
+                                   String caseDocApiUrl) {
+        this.restTemplate = restTemplate;
+        this.authTokenGenerator = authTokenGenerator;
+        this.caseDocApiUrl = caseDocApiUrl;
+    }
+
+    /**
+     * Given a file to upload, this will upload the file to the CCD document API and give back a unique ULR to access
+     * the uploaded file.
+     *
+     * @param authToken  the caller's bearer token used to verify the caller
+     * @param caseTypeId the area the file belongs to e.g. ET_EnglandWales
+     * @param file       the file to be uploaded
+     * @return the URL of the document we have just uploaded
+     * @throws CaseDocumentException if a problem occurs whilst uploading the document via API
+     */
+    public URI uploadDocument(String authToken, String caseTypeId, MultipartFile file) throws CaseDocumentException {
+        DocumentUploadResponse response = attemptWithRetriesToUploadDocumentToCaseDocumentApi(
+            0, authToken, caseTypeId, file);
+
+        CaseDocument caseDocument = validateResponse(
+            Objects.requireNonNull(response), file.getOriginalFilename());
+
+        return getUriFromFile(caseDocument);
+    }
+
+    private URI getUriFromFile(CaseDocument caseDocument) {
+        if (caseDocument.getLinks() != null
+            && caseDocument.getLinks().get("self") != null
+            && caseDocument.getLinks().get("self").get("href") != null) {
+            return URI.create(caseDocument.getLinks().get("self").get("href"));
+        }
+        return URI.create("");
+    }
```

Comment:
Worth just switching this around so if any are null, then return the empty.

But I'm also surprised how we can accept a blank URL being returned?  Would have thought that should throw an exception instead?  Only that what use would uploading a file to the service be if you can't get a handle on it?

Reply:
Uploaded changes, also the blank url was caught by regex validation, but updated to throw error directly from this snippet